src/core/rewrite/plan.rs

```diff
@@ -152,6 +158,9 @@ pub struct RebasePlanBuilder<'repo> {
     /// `y`.
     initial_constraints: HashMap<NonZeroOid, HashSet<NonZeroOid>>,
 
+    /// FIXME
+    messages_to_apply: Option<HashMap<NonZeroOid, String>>,
```

Comment:
I suspect that I'm misunderstanding something, but I tried implementing this and have gotten stuck. The rewrite and rebase appear to proceed w/o error, but the graph is left incorrect, w/ both the old commit and new/reworded commit still visible and branch/head not updated. For example, from the first test case (`test_reword_head`):
```
# initial state
    :
    O 62fc20d2 (test1) create test1.txt
    |
    @ 96d1c37a (> master) create test2.txt

# expected state
    :
    O 62fc20d2 (test1) create test1.txt
    |
    @ b38039bd (> master) foo

# actual state I'm getting
    :
    O 62fc20d2 (test1) create test1.txt
    |\
    | o b38039bd foo
    |
    @ 96d1c37a (> master) create test2.txt
```

So, it's creating the correctly and putting it at the right place, but it's not moving `master` and updating HEAD.

FWIW I pushed a `wip` commit w/ my current work. Basically:
1. in `reword`: create new amended commits, give these to `builder.replace_commit()`
2. in `builder.build`: add `Pick` to the rebase plan as normal but use the replacement commit oid instead of the original oid, if a replacement exists
3. in `execute_rebase_plan`, just deal w/ the `Pick` normally.  In fact, this file no longer has any changes from `master`, as far as I can tell. 

I'm assuming I'm missing a step that I can dig out of another command, but I need to stop for the night for now. If you get to this before I do, do I need to explicitly call restack after the rebase or something? Or am I missing a step somewhere in `execute_rebase_plan`? Sorry to be a bother and to need so much support on this!

Reply:
@claytonrcarter Oops, I didn't realize that there would be more steps to make the replacement commits work. We just need to update the `rebase_in_memory` function to let it know that replaced commits should be considered "rewritten" commits. I pushed the following commit to the branch `reword` on your repo (which apparently I'm allowed to do!):

```diff
commit 5ca1782d48eadc0fbd31bad7f25e712251e7f2c1
Author: Waleed Khan <me@waleedkhan.name>
Date:   Tue Mar 22 20:37:08 2022

    wip: correctly rewrite replaced commits

diff --git a/src/core/rewrite/execute.rs b/src/core/rewrite/execute.rs
index 12a0d4c..8b84024 100644
--- a/src/core/rewrite/execute.rs
+++ b/src/core/rewrite/execute.rs
@@ -273,6 +273,7 @@ pub fn check_out_updated_head(
 }
 
 /// Information about a merge conflict that occurred while moving commits.
+#[derive(Debug)]
 pub struct MergeConflictInfo {
     /// The OID of the commit that, when moved, caused a conflict.
     pub commit_oid: NonZeroOid,
@@ -443,12 +444,15 @@ mod in_memory {
                     current_oid = *commit_oid;
                 }
 
-                RebaseCommand::Pick { commit_oid } => {
+                RebaseCommand::Pick {
+                    original_commit_oid,
+                    commit_to_apply_oid,
+                } => {
                     let current_commit = repo
                         .find_commit_or_fail(current_oid)
                         .wrap_err("Finding current commit")?;
                     let commit_to_apply = repo
-                        .find_commit_or_fail(*commit_oid)
+                        .find_commit_or_fail(*commit_to_apply_oid)
                         .wrap_err("Finding commit to apply")?;
                     i += 1;
 
@@ -461,11 +465,11 @@ mod in_memory {
 
                     if commit_to_apply.get_parent_count() > 1 {
                         warn!(
-                            ?commit_oid,
+                            ?commit_to_apply_oid,
                             "BUG: Merge commit should have been detected during planning phase"
                         );
                         return Ok(RebaseInMemoryResult::CannotRebaseMergeCommit {
-                            commit_oid: *commit_oid,
+                            commit_oid: *commit_to_apply_oid,
                         });
                     };
 
@@ -483,7 +487,7 @@ mod in_memory {
                         Ok(rebased_commit) => rebased_commit,
                         Err(CherryPickFastError::MergeConflict { conflicting_paths }) => {
                             return Ok(RebaseInMemoryResult::MergeConflict(MergeConflictInfo {
-                                commit_oid: *commit_oid,
+                                commit_oid: *commit_to_apply_oid,
                                 conflicting_paths,
                             }))
                         }
@@ -493,7 +497,7 @@ mod in_memory {
                     let commit_message = commit_message.to_str().ok_or_else(|| {
                         eyre::eyre!(
                             "Could not decode commit message for commit: {:?}",
-                            commit_oid
+                            commit_to_apply_oid
                         )
                     })?;
 
@@ -526,8 +530,8 @@ mod in_memory {
                         )?,
                     )?;
                     if rebased_commit.is_empty() {
-                        rewritten_oids.push((*commit_oid, MaybeZeroOid::Zero));
-                        maybe_set_skipped_head_new_oid(*commit_oid, current_oid);
+                        rewritten_oids.push((*original_commit_oid, MaybeZeroOid::Zero));
+                        maybe_set_skipped_head_new_oid(*original_commit_oid, current_oid);
 
                         writeln!(
                             effects.get_output_stream(),
@@ -537,8 +541,10 @@ mod in_memory {
                             commit_description
                         )?;
                     } else {
-                        rewritten_oids
-                            .push((*commit_oid, MaybeZeroOid::NonZero(rebased_commit_oid)));
+                        rewritten_oids.push((
+                            *original_commit_oid,
+                            MaybeZeroOid::NonZero(rebased_commit_oid),
+                        ));
                         current_oid = rebased_commit_oid;
 
                         writeln!(
@@ -954,6 +960,7 @@ pub struct ExecuteRebasePlanOptions<'a> {
 
 /// The result of executing a rebase plan.
 #[must_use]
+#[derive(Debug)]
 pub enum ExecuteRebasePlanResult {
     /// The rebase operation succeeded.
     Succeeded,
diff --git a/src/core/rewrite/plan.rs b/src/core/rewrite/plan.rs
index a3534a6..8233d14 100644
--- a/src/core/rewrite/plan.rs
+++ b/src/core/rewrite/plan.rs
@@ -44,7 +44,10 @@ pub enum RebaseCommand {
 
     /// Apply the provided commit on top of the rebase head, and update the
     /// rebase head to point to the newly-applied commit.
-    Pick { commit_oid: NonZeroOid },
+    Pick {
+        original_commit_oid: NonZeroOid,
+        commit_to_apply_oid: NonZeroOid,
+    },
 
     Merge {
         /// The original merge commit to copy the commit message from.
@@ -83,7 +86,10 @@ impl ToString for RebaseCommand {
         match self {
             RebaseCommand::CreateLabel { label_name } => format!("label {}", label_name),
             RebaseCommand::Reset { target } => format!("reset {}", target.to_string()),
-            RebaseCommand::Pick { commit_oid } => format!("pick {}", commit_oid),
+            RebaseCommand::Pick {
+                original_commit_oid: _,
+                commit_to_apply_oid: commit_oid,
+            } => format!("pick {}", commit_oid),
             RebaseCommand::Merge {
                 commit_oid,
                 commits_to_merge,
@@ -346,12 +352,15 @@ impl<'repo> RebasePlanBuilder<'repo> {
             } else {
                 // Normal one-parent commit (or a zero-parent commit?), just
                 // rebase it and continue.
-                let commit_oid = current_commit.get_oid();
-                let commit_oid = match self.replacement_commits.get(&commit_oid) {
+                let original_commit_oid = current_commit.get_oid();
+                let commit_oid = match self.replacement_commits.get(&original_commit_oid) {
                     Some(replacement_oid) => *replacement_oid,
-                    None => commit_oid,
+                    None => original_commit_oid,
                 };
-                acc.push(RebaseCommand::Pick { commit_oid });
+                acc.push(RebaseCommand::Pick {
+                    original_commit_oid,
+                    commit_to_apply_oid: commit_oid,
+                });
                 acc.push(RebaseCommand::DetectEmptyCommit {
                     commit_oid: current_commit.get_oid(),
                 });
@@ -714,7 +723,10 @@ impl<'repo> RebasePlanBuilder<'repo> {
                 | RebaseCommand::Reset { target: _ }
                 | RebaseCommand::RegisterExtraPostRewriteHook
                 | RebaseCommand::DetectEmptyCommit { commit_oid: _ } => None,
-                RebaseCommand::Pick { commit_oid }
+                RebaseCommand::Pick {
+                    original_commit_oid: _,
+                    commit_to_apply_oid: commit_oid,
+                }
                 | RebaseCommand::Merge {
                     commit_oid,
                     commits_to_merge: _,
diff --git a/tests/command/test_move.rs b/tests/command/test_move.rs
index dc69c0e..10a734c 100644
--- a/tests/command/test_move.rs
+++ b/tests/command/test_move.rs
@@ -67,15 +67,15 @@ fn test_move_stick() -> eyre::Result<()> {
                         ),
                     },
                     Pick {
-                        commit_oid: NonZeroOid(70deb1e28791d8e7dd5a1f0c871a51b91282562f),
-                        message: None,
+                        original_commit_oid: NonZeroOid(70deb1e28791d8e7dd5a1f0c871a51b91282562f),
+                        commit_to_apply_oid: NonZeroOid(70deb1e28791d8e7dd5a1f0c871a51b91282562f),
                     },
                     DetectEmptyCommit {
                         commit_oid: NonZeroOid(70deb1e28791d8e7dd5a1f0c871a51b91282562f),
                     },
                     Pick {
-                        commit_oid: NonZeroOid(355e173bf9c5d2efac2e451da0cdad3fb82b869a),
-                        message: None,
+                        original_commit_oid: NonZeroOid(355e173bf9c5d2efac2e451da0cdad3fb82b869a),
+                        commit_to_apply_oid: NonZeroOid(355e173bf9c5d2efac2e451da0cdad3fb82b869a),
                     },
                     DetectEmptyCommit {
                         commit_oid: NonZeroOid(355e173bf9c5d2efac2e451da0cdad3fb82b869a),
@@ -255,15 +255,15 @@ fn test_move_with_source_not_in_smartlog() -> eyre::Result<()> {
                             ),
                         },
                         Pick {
-                            commit_oid: NonZeroOid(70deb1e28791d8e7dd5a1f0c871a51b91282562f),
-                            message: None,
+                            original_commit_oid: NonZeroOid(70deb1e28791d8e7dd5a1f0c871a51b91282562f),
+                            commit_to_apply_oid: NonZeroOid(70deb1e28791d8e7dd5a1f0c871a51b91282562f),
                         },
                         DetectEmptyCommit {
                             commit_oid: NonZeroOid(70deb1e28791d8e7dd5a1f0c871a51b91282562f),
                         },
                         Pick {
-                            commit_oid: NonZeroOid(355e173bf9c5d2efac2e451da0cdad3fb82b869a),
-                            message: None,
+                            original_commit_oid: NonZeroOid(355e173bf9c5d2efac2e451da0cdad3fb82b869a),
+                            commit_to_apply_oid: NonZeroOid(355e173bf9c5d2efac2e451da0cdad3fb82b869a),
                         },
                         DetectEmptyCommit {
                             commit_oid: NonZeroOid(355e173bf9c5d2efac2e451da0cdad3fb82b869a),
@@ -362,8 +362,8 @@ fn test_move_merge_conflict() -> eyre::Result<()> {
                         ),
                     },
                     Pick {
-                        commit_oid: NonZeroOid(e85d25c772a05b5c73ea8ec43881c12bbf588848),
-                        message: None,
+                        original_commit_oid: NonZeroOid(e85d25c772a05b5c73ea8ec43881c12bbf588848),
+                        commit_to_apply_oid: NonZeroOid(e85d25c772a05b5c73ea8ec43881c12bbf588848),
                     },
                     DetectEmptyCommit {
                         commit_oid: NonZeroOid(e85d25c772a05b5c73ea8ec43881c12bbf588848),
@@ -438,15 +438,15 @@ fn test_move_base() -> eyre::Result<()> {
                         ),
                     },
                     Pick {
-                        commit_oid: NonZeroOid(96d1c37a3d4363611c49f7e52186e189a04c531f),
-                        message: None,
+                        original_commit_oid: NonZeroOid(96d1c37a3d4363611c49f7e52186e189a04c531f),
+                        commit_to_apply_oid: NonZeroOid(96d1c37a3d4363611c49f7e52186e189a04c531f),
                     },
                     DetectEmptyCommit {
                         commit_oid: NonZeroOid(96d1c37a3d4363611c49f7e52186e189a04c531f),
                     },
                     Pick {
-                        commit_oid: NonZeroOid(70deb1e28791d8e7dd5a1f0c871a51b91282562f),
-                        message: None,
+                        original_commit_oid: NonZeroOid(70deb1e28791d8e7dd5a1f0c871a51b91282562f),
+                        commit_to_apply_oid: NonZeroOid(70deb1e28791d8e7dd5a1f0c871a51b91282562f),
                     },
                     DetectEmptyCommit {
                         commit_oid: NonZeroOid(70deb1e28791d8e7dd5a1f0c871a51b91282562f),
@@ -571,8 +571,8 @@ fn test_move_checkout_new_head() -> eyre::Result<()> {
                         ),
                     },
                     Pick {
-                        commit_oid: NonZeroOid(fe65c1fe15584744e649b2c79d4cf9b0d878f92e),
-                        message: None,
+                        original_commit_oid: NonZeroOid(fe65c1fe15584744e649b2c79d4cf9b0d878f92e),
+                        commit_to_apply_oid: NonZeroOid(fe65c1fe15584744e649b2c79d4cf9b0d878f92e),
                     },
                     DetectEmptyCommit {
                         commit_oid: NonZeroOid(fe65c1fe15584744e649b2c79d4cf9b0d878f92e),
@@ -642,8 +642,8 @@ fn test_move_branch() -> eyre::Result<()> {
                         ),
                     },
                     Pick {
-                        commit_oid: NonZeroOid(98b9119d16974f372e76cb64a3b77c528fc0b18b),
-                        message: None,
+                        original_commit_oid: NonZeroOid(98b9119d16974f372e76cb64a3b77c528fc0b18b),
+                        commit_to_apply_oid: NonZeroOid(98b9119d16974f372e76cb64a3b77c528fc0b18b),
                     },
                     DetectEmptyCommit {
                         commit_oid: NonZeroOid(98b9119d16974f372e76cb64a3b77c528fc0b18b),
@@ -832,8 +832,8 @@ fn test_move_in_memory_gc() -> eyre::Result<()> {
                         ),
                     },
                     Pick {
-                        commit_oid: NonZeroOid(96d1c37a3d4363611c49f7e52186e189a04c531f),
-                        message: None,
+                        original_commit_oid: NonZeroOid(96d1c37a3d4363611c49f7e52186e189a04c531f),
+                        commit_to_apply_oid: NonZeroOid(96d1c37a3d4363611c49f7e52186e189a04c531f),
                     },
                     DetectEmptyCommit {
                         commit_oid: NonZeroOid(96d1c37a3d4363611c49f7e52186e189a04c531f),
@@ -1632,8 +1632,8 @@ fn test_move_merge_commit() -> eyre::Result<()> {
                             ),
                         },
                         Pick {
-                            commit_oid: NonZeroOid(fe65c1fe15584744e649b2c79d4cf9b0d878f92e),
-                            message: None,
+                            original_commit_oid: NonZeroOid(fe65c1fe15584744e649b2c79d4cf9b0d878f92e),
+                            commit_to_apply_oid: NonZeroOid(fe65c1fe15584744e649b2c79d4cf9b0d878f92e),
                         },
                         DetectEmptyCommit {
                             commit_oid: NonZeroOid(fe65c1fe15584744e649b2c79d4cf9b0d878f92e),
```

After this change, `cargo test reword` passes.

The above patch isn't sufficient to get it to work for on-disk rebases, but don't worry about that. I'll fix it myself when (if?) we have a reason to use `reword` with on-disk rebases.

> Sorry to be a bother and to need so much support on this!

It's no trouble at all, I'm excited for this feature to land! 😃